File: backend/server/src/main/java/com/groudina/ten/demo/jwt/SecurityContextRepository.java
package com.groudina.ten.demo.jwt;

import lombok.extern.log4j.Log4j2;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpHeaders;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContext;
import org.springframework.security.core.context.SecurityContextImpl;
import org.springframework.security.web.server.context.ServerSecurityContextRepository;
import org.springframework.stereotype.Component;
import org.springframework.web.server.ServerWebExchange;
import reactor.core.publisher.Mono;

@Log4j2
@Component
public class SecurityContextRepository implements ServerSecurityContextRepository {
    private static final String TOKEN_PREFIX = "Bearer ";

    @Autowired
    private AuthenticationManager authenticationManager;

    @Override
    public Mono<Void> save(ServerWebExchange serverWebExchange, SecurityContext securityContext) {
        throw new UnsupportedOperationException("Not supported yet.");
    }

    @Override
    public Mono<SecurityContext> load(ServerWebExchange serverWebExchange) {
        String authHeader = serverWebExchange.getRequest().getHeaders().getFirst(HttpHeaders.AUTHORIZATION);

        if (authHeader != null && authHeader.startsWith(TOKEN_PREFIX)) {
            String authToken = authHeader.replace(TOKEN_PREFIX, "");
            Authentication auth = new UsernamePasswordAuthenticationToken(authToken, authToken);

            return authenticationManager.authenticate(auth).map(SecurityContextImpl::new);
        } else {
            log.warn("There is no 'Bearer ' auth header");
            return Mono.empty();
        }
    }
}

File: backend/app/web/src/main/kotlin/ru/nekoguys/game/web/security/SecurityContextRepository.kt
package ru.nekoguys.game.web.security

import org.slf4j.LoggerFactory
import org.springframework.http.HttpHeaders
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.core.context.SecurityContext
import org.springframework.security.core.context.SecurityContextImpl
import org.springframework.security.web.server.context.ServerSecurityContextRepository
import org.springframework.web.server.ServerWebExchange
import reactor.core.publisher.Mono

class SecurityContextRepository (
    private val authenticationManager: GameAuthenticationManager,
) : ServerSecurityContextRepository {

    override fun save(serverWebExchange: ServerWebExchange, securityContext: SecurityContext): Mono<Void> {
        throw UnsupportedOperationException("Not supported yet.")
    }

    override fun load(serverWebExchange: ServerWebExchange): Mono<SecurityContext> {
        val authHeader: String? =
            serverWebExchange
                .request
                .headers
                .getFirst(HttpHeaders.AUTHORIZATION)

        return if (authHeader != null && authHeader.startsWith(TOKEN_PREFIX)) {
            val authToken = authHeader.replace(TOKEN_PREFIX, "")
            createContextFromToken(authToken)
        } else {
            logger.warn("There is no 'Bearer ' auth header")
            Mono.empty()
        }
    }

    private fun createContextFromToken(authToken: String): Mono<SecurityContext> {
        val auth = UsernamePasswordAuthenticationToken(authToken, authToken)
        return authenticationManager
            .authenticate(auth)
            .map(::SecurityContextImpl)
    }

    companion object {
        private const val TOKEN_PREFIX = "Bearer "
        private val logger = LoggerFactory.getLogger(SecurityContextRepository::class.java)
    }
}
